File: app/src/main/java/io/github/zwieback/familyfinance/business/account/lifecycle/creator/AccountCreator.java
package io.github.zwieback.familyfinance.business.account.lifecycle.creator;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.math.BigDecimal;
import java.util.Set;
import java.util.TreeSet;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.core.lifecycle.creator.EntityCreator;
import io.github.zwieback.familyfinance.core.model.Account;
import io.github.zwieback.familyfinance.core.model.Currency;
import io.github.zwieback.familyfinance.core.model.Person;
import io.requery.Persistable;
import io.requery.reactivex.ReactiveEntityStore;
import io.requery.reactivex.ReactiveResult;

public class AccountCreator extends EntityCreator<Account> {

    public AccountCreator(Context context, ReactiveEntityStore<Persistable> data) {
        super(context, data);
    }

    @Override
    protected Iterable<Account> buildEntities() {
        Currency defaultCurrency = findCurrency(databasePrefs.getCurrencyId());
        Person chief = findPerson(databasePrefs.getPersonId());

        Set<Account> accounts = new TreeSet<>(this);
        Account cash = createAccount(null, defaultCurrency, chief, getString(R.string.account_cash),
                false, BigDecimal.ZERO, 1);
        accounts.add(cash);
        return accounts;
    }

    @Override
    public int compare(Account left, Account right) {
        return Integer.valueOf(left.getOrderCode()).compareTo(right.getOrderCode());
    }

    @NonNull
    private Person findPerson(int personId) {
        ReactiveResult<Person> persons = data
                .select(Person.class)
                .where(Person.ID.eq(personId))
                .get();
        return persons.first();
    }

    @NonNull
    private Currency findCurrency(int currencyId) {
        ReactiveResult<Currency> currencies = data
                .select(Currency.class)
                .where(Currency.ID.eq(currencyId))
                .get();
        return currencies.first();
    }

    private static Account createAccount(@Nullable Account parent,
                                         @Nullable Currency currency,
                                         @Nullable Person owner,
                                         @NonNull String name,
                                         boolean folder,
                                         @Nullable BigDecimal initialBalance,
                                         int orderCode) {
        return new Account()
                .setActive(true)
                .setParent(parent)
                .setCurrency(currency)
                .setOwner(owner)
                .setName(name)
                .setFolder(folder)
                .setInitialBalance(initialBalance)
                .setOrderCode(orderCode);
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/account/lifecycle/creator/AccountCreator.kt
package io.github.zwieback.familyfinance.business.account.lifecycle.creator

import android.content.Context
import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.core.lifecycle.creator.EntityCreator
import io.github.zwieback.familyfinance.core.model.Account
import io.github.zwieback.familyfinance.core.model.Currency
import io.github.zwieback.familyfinance.core.model.Person
import io.github.zwieback.familyfinance.core.model.type.AccountType
import io.requery.Persistable
import io.requery.reactivex.ReactiveEntityStore
import java.math.BigDecimal

class AccountCreator(
    context: Context,
    data: ReactiveEntityStore<Persistable>
) : EntityCreator<Account>(context, data) {

    override fun buildEntities(): Iterable<Account> {
        val defaultCurrency = findCurrency(databasePrefs.currencyId)
        val chief = findPerson(databasePrefs.personId)
        val cash = createAccount(
            null,
            defaultCurrency,
            chief,
            getString(R.string.account_cash),
            false,
            BigDecimal.ZERO,
            1,
            AccountType.CASH_ACCOUNT
        )
        return sortedSetOf(this, cash)
    }

    override fun compare(left: Account, right: Account): Int {
        return left.orderCode.compareTo(right.orderCode)
    }

    private fun findPerson(personId: Int): Person {
        val persons = data
            .select(Person::class.java)
            .where(Person.ID.eq(personId))
            .get()
        return persons.first()
    }

    private fun findCurrency(currencyId: Int): Currency {
        val currencies = data
            .select(Currency::class.java)
            .where(Currency.ID.eq(currencyId))
            .get()
        return currencies.first()
    }

    private fun createAccount(
        parent: Account?,
        currency: Currency?,
        owner: Person?,
        name: String,
        folder: Boolean,
        initialBalance: BigDecimal?,
        orderCode: Int,
        type: AccountType
    ): Account {
        return Account()
            .setActive(true)
            .setParent(parent)
            .setCurrency(currency)
            .setOwner(owner)
            .setName(name)
            .setFolder(folder)
            .setInitialBalance(initialBalance)
            .setOrderCode(orderCode)
            .setType(type)
    }
}
